source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -55,6 +59,17 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   }
   cluster_info_ = cluster->info();
 
+  // We only need to do DNS lookups for hosts in dynamic forward proxy clusters,
+  // since the other cluster types do their own DNS management.
+  const absl::optional<CustomClusterType>& cluster_type = cluster_info_->clusterType();
```

Comment:
I thought about it - it was already subject to this problem as we only did DNS checks once when receiving the headers, so either before or after this change, if your rewrite and route change happened upstream of this filter, you would have the same problem.

Honestly I think there's a bunch of code which could benefit from an onRouteChanged function, but I think given that issue was pre-existing it isn't in the scope of this PR.   WDYT?

Reply:
> Honestly I think there's a bunch of code which could benefit from an onRouteChanged function, but I think given that issue was pre-existing it isn't in the scope of this PR. WDYT?

Agreed, this would be a nice addition but seems out of scope.